CONTRIBUTING.md

```diff
@@ -1,52 +1,120 @@
+- Consider how it fits with Hanami framework design, and how it might impact existing features or workflows.
+- We'll discuss discuss your proposal with you if/how it should be built.
 
-3. Create a new branch and make your changes. This includes tests for features!
+For smaller changes that are more easily demonstrated through code, feel free to [submit a pull request](#submitting-pull-requests).
 
-4. Push to your fork and submit a pull request. For more information, see
-[GitHub's pull request help section](https://help.github.com/articles/using-pull-requests/).
+We cannot add all suggested enhancements to Hanami, but we’re open to having constructive conversations with you about them, and learning from each other along the way.
 
-At this point you're waiting on us. Expect a conversation regarding your pull
-request; Questions, clarifications, and so on.
+### Submitting pull requests
 
-Some things that will increase the chance that your pull request is accepted:
+When you submit a pull request, be sure to consider the following:
 
-* Use Hanami idioms
-* Include tests that fail without your code, and pass with it
-* Update the documentation, guides, etc.
+- Include tests for your change.
+- Ensure tests are passing on CI (see the “Actions” tab for each repo).
+- Add or update API documentation for your change.
+- Update user guides and/or docs to reflect your change, via a pull request to [hanami/site](https://github.com/hanami/site).
+- Add an entry to [CHANGELOG.md](./CHANGELOG.md) for your change.
+- Follow the [conventions](#conventions) outlined below, plus any other idioms you see across the codebase.
+
+If you need help with any of these, please ask! Once your pull request is ready, we’ll review your change and have a conversation with you about it.
+
+## Development setup
+
+After cloning the repo:
+
+```bash
+bundle install
+npm install
+```
+
+To run the rests:
+
+```bash
+bundle exec rake
+# or
+bundle exec rspec
+```
+
+## Conventions
+
+### Commit messages
+
+Your pull request will be squashed into a single commit when merged. Break up your pull request into individual commits if it will help with review of the code.
+
+One of your commits should contain what the message that will become the commit message for the merged pull request.
+
+[Follow these guidelines](https://developer.vonage.com/en/blog/how-to-write-a-great-git-commit-message) for your commit messages. If in doubt, don’t worry too much, we can fix things when we merge your pull request.
+
+### Changelog
+
+When you’re making a change that Hanami users should know about, please update the `CHANGELOG.md`. We follow the [keep a changelog](https://keepachangelog.com) conventions for our changelogs.
+
+New entries can go in the `## [Unreleased]` section at the top of the file. For example:
+
+```md
+## [Unreleased]
+
+### Added
+
+- Add "418 I'm a teapot" middleware for caffeine allergics (@jane_doe in #9999)
```

Comment:
FYI, with this change, I've intentionally changed the CHANGELOG format so it doesn't start with `[Contributor Name]`, which I never really liked.

I'll aim to get our existing CHANGELOG files updated to match this format in the future. In the meantime, I don't think it hurts to prescribe our desired format here :)